src/docfx/build/toc/TableOfContentsMap.cs

```diff
@@ -53,54 +57,70 @@ public string FindTocRelativePath(Document file)
         /// </summary>
         public Document GetNearestToc(Document file)
         {
-            // fallback to all tocs if no toc files reference this file
-            var hasReferencedToc = _documentToTocs.TryGetValue(file, out var referencedTocFiles);
-            var filteredTocFiles = hasReferencedToc ? (IEnumerable<Document>)referencedTocFiles : _tocs;
-
             var nearestToc = (Document)null;
-            var nearestSubDirCount = 0;
-            var nearestParentDirCount = 0;
-            foreach (var toc in filteredTocFiles)
-            {
-                var relativePath = PathUtility.GetRelativePathToFile(file.SitePath, toc.SitePath);
-                var (subDirCount, parentDirCount) = GetDirectoryCount(relativePath);
-                if (!hasReferencedToc && subDirCount != 0)
-                {
-                    continue;
-                }
-                var distance = Compare(nearestSubDirCount, nearestParentDirCount, subDirCount, parentDirCount);
-                if (nearestToc == null || distance > 0 ||
-                    (distance == 0 && string.Compare(nearestToc.SitePath, toc.SitePath, PathUtility.PathComparison) > 0))
-                {
-                    nearestToc = toc;
-                    nearestSubDirCount = subDirCount;
-                    nearestParentDirCount = parentDirCount;
-                }
-            }
-
-            return nearestToc;
-        }
 
-        private static int Compare(int xSubDirCount, int xParentDirCount, int ySubDirCount, int yParentDirCount)
-        {
-            if (xSubDirCount == ySubDirCount)
+            if (!_documentToTocs.TryGetValue(file, out var referencedTocFiles))
             {
-                return xParentDirCount - yParentDirCount;
+                // fallback to all tocs if no toc files reference this file
+                // filter toc files by relative path: subdircount = 0, min parentdir
+                // get default nonreferenced toc
+                nearestToc = (from toc in _tocs
+                              let dirInfo = GetRelativeDirectoryInfo(file, toc)
+                              where dirInfo.subDirectoryCount == 0
+                              orderby dirInfo.parentDirectoryCount
+                              select toc)
+                             .FirstOrDefault();
             }
             else
             {
-                return xSubDirCount - ySubDirCount;
+                // from referenced pick the nearest one
+                // 1. sub count
+                // 2. sub nearest.
+                // 3. parent nearest
+                // 4. sub-name word-level levenshtein distance nearest
+                // 5. sub-name lexicographical nearest
+                var tocCandidates = (from toc in referencedTocFiles
+                                     let dirInfo = GetRelativeDirectoryInfo(file, toc)
+                                     orderby dirInfo.subDirectoryCount,
```

Comment:
>orderby [](start = 37, length = 7)

we only want to top ones, no need to order all

Reply:
this is a great lesson